### PR TITLE
Allow Rectangle and Ellipse selectors to be rotated

### PR DESCRIPTION
**Pull request by dstansby**
## PR Summary
This PR allows `RectangleSelector` and `EllipseSelector` to be rotated about their anchor point. This is enabled by default with the `'r'` modifier key. Example:

https://user-images.githubusercontent.com/6197628/113521328-ad48f100-9590-11eb-9ee5-e7b539b99ff6.mov

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] New features are documented, with examples if plot related.
- [ x Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -2418,25 +2435,43 @@ def _release(self, event):
...
-               (self.drag_from_anywhere and self._contains(event))) and
-              self._extents_on_press is not None):
-            x0, x1, y0, y1 = self._extents_on_press
-            dx = event.xdata - self.eventpress.xdata
-            dy = event.ydata - self.eventpress.ydata
-            x0 += dx
-            x1 += dx
-            y0 += dy
-            y1 += dy
+        elif (('move' in self.state or self.active_handle == 'C')):
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2453]:
```suggestion
        elif 'move' in self.state or self.active_handle == 'C':
```

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -2418,25 +2435,43 @@ def _release(self, event):
...
+        if 'rotate' in self.state:
+            rotation_new = np.arctan2(ymove - y0,
+                                      xmove - x0)
+            rotation_old = np.arctan2(self.prev_ymove - y0,
+                                      self.prev_xmove - x0)
+            rotation += rotation_new - rotation_old
 
         # move existing shape
-        elif (('move' in self.state or self.active_handle == 'C' or
-               (self.drag_from_anywhere and self._contains(event))) and
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2431]: What happened to `drag_from_anywhere` support?

File: lib/matplotlib/widgets.py [hunk 3]
```diff
@@ -2529,26 +2591,36 @@ def extents(self, extents):
...
         ylim = sorted(self.ax.get_ylim())
 
-        xmin = max(xlim[0], xmin)
-        ymin = max(ylim[0], ymin)
-        xmax = min(xmax, xlim[1])
-        ymax = min(ymax, ylim[1])
+        x0 = np.clip(x0, *xlim)
+        x1 = np.clip(x1, *xlim)
+        y0 = np.clip(y0, *ylim)
+        y1 = np.clip(y1, *ylim)
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2614]: This doesn't seem right, as a rotated rectangle can be wider than the x limits, but these extents are for the un-rotated rectangle..

File: lib/matplotlib/widgets.py [hunk 4]
```diff
@@ -2580,14 +2652,7 @@ def _set_active_handle(self, event):
...
 
-        # Save coordinates of rectangle at the start of handle movement.
-        x0, x1, y0, y1 = self.extents
-        # Switch variables so that only x1 and/or y1 are updated on move.
-        if self.active_handle in ['W', 'SW', 'NW']:
-            x0, x1 = x1, event.xdata
-        if self.active_handle in ['N', 'NW', 'NE']:
-            y0, y1 = y1, event.ydata
-        self._extents_on_press = x0, x1, y0, y1
+        x0, y0, width, height, angle = self._rect_properties
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2655]: You don't seem to be doing anything with this result.

**Review by QuLogic — COMMENTED** [review]
There should be tests for resizing after the selector has rotated, which is a thing that could not occur before.

File: lib/matplotlib/tests/test_widgets.py [hunk 5]
```diff
@@ -189,6 +189,32 @@ def onselect(epress, erelease):
...
 
 
+def test_rectangle_rotate():
+    ax = get_ax()
+
+    def onselect(epress, erelease):
+        pass
+
+    tool = widgets.RectangleSelector(ax, onselect=onselect, interactive=True,
+                                     state_modifier_keys={'rotate': 'r'})
```
[the remark this answers is not in the document]
Comment (dstansby) [reply lib/matplotlib/tests/test_widgets.py:199]: Ah yes, this is a hangover from before I made it default.

File: lib/matplotlib/widgets.py (hunk 3 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py:2614]: I'm not sure why these lines were here in the first place, so I've just removed this bit.

File: lib/matplotlib/widgets.py (hunk 2 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py:2431]: Woops, I think this got lost in the rebase I had to do.

File: lib/matplotlib/patches.py [hunk 6]
```diff
@@ -779,10 +779,22 @@ def get_patch_transform(self):
         # important to call the accessor method and not directly access the
         # transformation member variable.
         bbox = self.get_bbox()
-        return (transforms.BboxTransformTo(bbox)
-                + transforms.Affine2D().rotate_deg_around(
+        return (transforms.BboxTransformTo(bbox) +
+                transforms.Affine2D().rotate_deg_around(
                     bbox.x0, bbox.y0, self.angle))
 
+    def _get_translate_rotate_transform(self):
```
[the remark this answers is not in the document]
Comment (dstansby) [reply lib/matplotlib/patches.py]: 👍 , I've moved it to the `RectangleSelector` class.

File: lib/matplotlib/widgets.py (hunk 2 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py:2430]: Woops, got lost in a rebase, should be back now.

**Comment by dstansby** [comment]
Thanks for the review - I marked all the minor changes as resolved, and have addressed the other points too.

File: lib/matplotlib/widgets.py [hunk 7]
```diff
@@ -2370,6 +2394,8 @@ def _press(self, event):
             y = event.ydata
             self.extents = x, x, y, y
 
+        self.prev_xmove = event.xdata
+        self.prev_ymove = event.ydata
```
Comment (QuLogic) [inline lib/matplotlib/widgets.py:2398]: No need for this to be public?
```suggestion
        self._prev_xmove = event.xdata
        self._prev_ymove = event.ydata
```

**Review by QuLogic — COMMENTED** [review]
Still needs a test of resizing after rotation.

File: lib/matplotlib/widgets.py (hunk 7 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py:2398]: Nope, I've made it private

**Comment by dstansby** [comment]
> Still needs a test of resizing after rotation.

Thanks for catching that, it revealed a (now fixed) bug in how I was calculating the resizing. I've also paramatrized the test so it runs for both `Rectangle` and `Ellipse` selectors.

File: lib/matplotlib/widgets.py [hunk 8]
```diff
@@ -2325,12 +2333,12 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
...
             props = dict(markeredgecolor=rectprops.get('edgecolor', 'r'))
         props.update(cbook.normalize_kwargs(marker_props, Line2D._alias_map))
-        self._corner_order = ['NW', 'NE', 'SE', 'SW']
+        self._corner_order = ['SW', 'SE', 'NE', 'NW']
         xc, yc = self.corners
         self._corner_handles = ToolHandles(self.ax, xc, yc, marker_props=props,
                                            useblit=self.useblit)
 
-        self._edge_order = ['W', 'N', 'E', 'S']
+        self._edge_order = ['W', 'S', 'E', 'N']
```
Comment (jklymak) [inline lib/matplotlib/widgets.py:2341]: This order hurts my brain ;-)  Cardinal directions should be clockwise and start w/ north!.  But, I don't think it really matters here....

File: lib/matplotlib/widgets.py [hunk 9]
```diff
@@ -2350,7 +2358,23 @@ def onselect(eclick: MouseEvent, erelease: MouseEvent)
...
+
+    def _get_translate_rotate_transform(self):
+        """
+        Return a transform that translates the LH corner to the origin,
+        and de-rotates the rectangle.
+        """
+        rect = self.to_draw
+        x0, y0, width, height, angle = self._rect_properties
+        return (transforms.Affine2D()
+                .rotate_deg_around(x0, y0, -np.rad2deg(angle))
```
Comment (jklymak) [inline lib/matplotlib/widgets.py:2376]:
```suggestion
                .rotate_around(x0, y0, -angle)
```
??

File: lib/matplotlib/widgets.py [hunk 10]
```diff
@@ -2468,57 +2522,84 @@ def _onmove(self, event):
...
     def corners(self):
         """Corners of rectangle from lower left, moving clockwise."""
-        x0, y0, width, height = self._rect_bbox
-        xc = x0, x0 + width, x0 + width, x0
-        yc = y0, y0, y0 + height, y0 + height
-        return xc, yc
+        x0, y0, width, height, angle = self._rect_properties
+        x = np.array([0, width, width, 0])
+        y = np.array([0, 0, height, height])
+        xc = x0 + (np.cos(angle) * x - np.sin(angle) * y)
```
Comment (jklymak) [inline lib/matplotlib/widgets.py:2559]: You have a ton of these rotations.  Is there a reason you don't use `mtransforms.rotate` for these?

File: lib/matplotlib/widgets.py (hunk 9 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py:2376]: 👍

File: lib/matplotlib/widgets.py (hunk 10 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py]: Nope, in fact it turns out this is just the inverse of the transform I added a method to get, so these are all much simplified now, thanks for catching.

**Comment by QuLogic** [comment]
OK, I tried this out with the rectangle selector example, and it does not work at all in a way I'd expect. One edge barely rotated, while the other turns much more. Sometimes, the rectangle is a completely different aspect ratio.

**Comment by dstansby** [comment]
That's because the rectangle is defined in axes coordinates, not figure coordinates. If you do `ax.set_aspect('equal')` it should work as expected. I think it would take quite a bit more work to convert `RectangleSelector` to be drawn in figure coordinates instead of axes coordinates, but happy to give it a go if people think it's worth blocking this PR on.

**Comment by jklymak** [comment]
Yeah I was just getting the point of wondering about that.  Is it defined in axes coordinates or data coordinates?   If data coordinates, I'm not clear how this is useful.

**Comment by dstansby** [comment]
I've had a bit of a think about this, and would like to advocate for it to be included as is, with a note (or warning) in the docstring that it doesn't work as expected when the axes aspect ratio != 1. My reasoning is:

- It doesn't break or modify any existing behavior
- The motivating use case for this PR was selecting regions in astronomical images. These are almost always displayed with an aspect ratio of 1, where rotating selectors works.

> Is it defined in axes coordinates or data coordinates?

It is defined in data coordinates, sorry for the confusion in my earlier comment.

**Comment by jklymak** [comment]
@dstansby I understand your use case.  OTOH I think its going to be confusing for everyone else when this rotation is not in figure/axes space...  Is it hard to put in figure space?

**Comment by jklymak** [comment]
@dstansby did you want a definitive answer on this?  We could discuss on the call.  I guess I'm -0.25 the way the PR is now.  Can you implement outside of core?

**Comment by dstansby** [comment]
>  Is it hard to put in figure space?

I suspect the answer is technically not hard, but it will take a reasonable amount of development time to disentangle everything to maintain backwards compatibility

> Can you implement outside of core?

What do you mean by this? In another package?

I think the deciding question is: Would it be okay to release this as is, and then in the future change (ie. fix) the behavior for aspect!=1 without warning? If not, I'm happy to take a dive into moving the widgets from data to axes coordinates before merging this, but it might take a while.

**Comment by jklymak** [comment]
I personally think this should be in figure space, but wouldn't block over it.

**Comment by timhoffm** [comment]
We should not inroduce broken rotations. That's rather confusing.

If you cannot make this work for the general case right now, I propose to only enable rotations on equal-aspect Axes. This should be easy to check via `self.ax.get_aspect() == 1`.

**Comment by dstansby** [comment]
> If you cannot make this work for the general case right now, I propose to only enable rotations on equal-aspect Axes. This should be easy to check via self.ax.get_aspect() == 1.

Sounds good to me 👍 

For a non-equal aspect axes, trying to rotate now does nothing and raises a warning. I've added an extra test to check this works as intended, and updated the docs.

File: lib/matplotlib/widgets.py [hunk 11]
```diff
@@ -1823,9 +1825,11 @@ def press(self, event):
             self._prev_event = event
             key = event.key or ''
             key = key.replace('ctrl', 'control')
-            # move state is locked in on a button press
-            if key == self.state_modifier_keys['move']:
-                self.state.add('move')
+            # move/rotate state is locked in on a button press
+            for action_key in ['move', 'rotate']:
```
Comment (timhoffm) [inline lib/matplotlib/widgets.py:1829]: Shouldn't this rather be named
```suggestion
            for action in ['move', 'rotate']:
```
?

**Review by timhoffm — COMMENTED** [review]
Is it possible with reasonable effort to rotate about the center? The tilting over an edge is quite surprising and I don't know any other application doing this.

File: lib/matplotlib/widgets.py (hunk 11 above)
Comment (dstansby) [reply lib/matplotlib/widgets.py]: Yes, this is definitely better to avoid confusion between dictionary and keyboard keys 😄 - changed.

**Comment by dstansby** [comment]
I've rebased this, so should be good for re-review.

**Comment by timhoffm** [comment]
> I just spent 5 minutes trying to work out if there was an easy way, and couldn't work it out, so I think changing this would take a reasonable amount of effort. This is mainly because a `Rectangle` patch is defined by a corner coordinate and rotation, so it's non trivial (but possible) if one wants to set the rotation angle to be about the center, as a new rotation angle about the corner and corner coordinate has to be calculated for the rectangle at the same time.

Isn't the new location simply: `x' = x + d - Md` where x is the position vector of the rectangle corner, d is the connection vector from the corner to the center, and M is the rotation matrix?

**Comment by dstansby** [comment]
> Isn't the new location simply: `x' = x + d - Md` where x is the position vector of the rectangle corner, d is the connection vector from the corner to the center, and M is the rotation matrix?

Possibly? One also needs to work out the new rotation (about the patch corner). If someone wants to take a go at this feel free, I don't currently have the bandwidth to work out the maths or where it needs changing in the code.

**Comment by timhoffm** [comment]
> One also needs to work out the new rotation (about the patch corner). 
I think, you can just do the rotation and then shift the rectangle as proposed.

**Comment by ericpre** [comment]
#20839 is an alternative PR to add selector rotation.

**Comment by ericpre** [comment]
Done in #20839.

**Pull request closed without merging**